In [2]:
import numpy as np
import random

## Sort

In [243]:
ls_s = [2,3,4,1,20,15]
ls_b = [i for i in range(10**3)]
np.random.shuffle(ls_b)

In [244]:
del ls_b

### Selecetion Sort
- Computation Complexity: O(n^2)
- Space Complexity: O(1)

In [65]:
def selection_sort(ls):
    for i in range(len(ls)-1):
        smallest = i
        for j in range(i+1,len(ls)):
            if ls[j]<ls[smallest]: smallest = j
        ls[i],ls[smallest] = ls[smallest],ls[i]
    return ls

In [66]:
%%time
selection_sort(ls_s)

Wall time: 0 ns


[1, 2, 3, 4, 15, 20]

In [67]:
%%time
selection_sort(ls_b)

Wall time: 37.9 ms


[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


### Merge Sort
- A type of Divide and Conqure
- Computation Complexity: O(n * log(n))

In [54]:
def merge_sort(ls):
    if len(ls)<=1: return ls
    if len(ls)==2: 
        if ls[0]<ls[1]: return ls
        else: return ls[::-1]
        
    mid = len(ls) // 2
    return merge(merge_sort(ls[:mid]), merge_sort(ls[mid:]))

In [57]:
%%time
merge_sort(ls_s)

Wall time: 0 ns


[1, 2, 3, 4, 15, 20]

In [58]:
%%time
merge_sort(ls_b)

Wall time: 2.99 ms


[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


### Quick Sort - No implementation made
- Computation Complexity: O(n*log(n))

## Merging Two Sorted List
- Computation Complexity: O(n)
- Space Complexity: O(n)

Tried two algos: one is using list and the other using stack (list but using as stack). Appearently, the latter is faster.

In [61]:
ls1_s = [1,5,10]; ls2_s = [2,3,9]
ls12_b = [i for i in range(2*10**3)]
np.random.shuffle(ls12_b)
ls1_b = ls12_b[0:10**3]; ls2_b = ls12_b[10**3:2*10**3]
ls1_b.sort(); ls2_b.sort()

In [79]:
del ls1_b
del ls2_b
del ls12_b

In [59]:
def merge(ls1, ls2):
    res = []
    ls1r, ls2r = ls1[::-1], ls2[::-1]
    while (ls1r!=[] and ls2r!=[]):
        if (ls1r[-1]<ls2r[-1]):
            res.append(ls1r[-1])
            ls1r.pop()
        else:
            res.append(ls2r[-1])
            ls2r.pop()
            
    if (ls1r!=[]): res.extend(ls1r[::-1])
    if (ls2r!=[]): res.extend(ls2r[::-1])
        
    return res

def merge_nonstack(ls1, ls2):
    res = []
    ls1r, ls2r = [ls1[i] for i in range(len(ls1))], [ls2[i] for i in range(len(ls2))]
    while (ls1r!=[] and ls2r!=[]):
        if (ls1r[0]<ls2r[0]):
            res.append(ls1r[0])
            ls1r = ls1r[1:]
        else:
            res.append(ls2r[0])
            ls2r = ls2r[1:]
    
    if (ls1r!=[]): res.extend(ls1r)
    if (ls2r!=[]): res.extend(ls2r)    
    
        
    return res

In [39]:
%%time
merge(ls1_s,ls2_s)

Wall time: 0 ns


[1, 2, 3, 5, 9]

In [63]:
%%time
merge(ls1_b,ls2_b)

Wall time: 996 µs


[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


In [62]:
%%time
merge_nonstack(ls1_b,ls2_b)

Wall time: 3.83 ms


[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


## Generate Power Set
- Computation Complex: O(2^n)

In [100]:
def power_set(ls):
    res = [[]]
    for item in ls:
        res.extend([subset + [item] for subset in res])
    return res

In [111]:
%%time
sorted(power_set(['a','b','c']),key=len)

Wall time: 0 ns


[[], ['a'], ['b'], ['c'], ['a', 'b'], ['a', 'c'], ['b', 'c'], ['a', 'b', 'c']]

In [108]:
%%time
power_set([str(i) for i in range(20)])

Wall time: 362 ms


[[],
 ['0'],
 ['1'],
 ['0', '1'],
 ['2'],
 ['0', '2'],
 ['1', '2'],
 ['0', '1', '2'],
 ['3'],
 ['0', '3'],
 ['1', '3'],
 ['0', '1', '3'],
 ['2', '3'],
 ['0', '2', '3'],
 ['1', '2', '3'],
 ['0', '1', '2', '3'],
 ['4'],
 ['0', '4'],
 ['1', '4'],
 ['0', '1', '4'],
 ['2', '4'],
 ['0', '2', '4'],
 ['1', '2', '4'],
 ['0', '1', '2', '4'],
 ['3', '4'],
 ['0', '3', '4'],
 ['1', '3', '4'],
 ['0', '1', '3', '4'],
 ['2', '3', '4'],
 ['0', '2', '3', '4'],
 ['1', '2', '3', '4'],
 ['0', '1', '2', '3', '4'],
 ['5'],
 ['0', '5'],
 ['1', '5'],
 ['0', '1', '5'],
 ['2', '5'],
 ['0', '2', '5'],
 ['1', '2', '5'],
 ['0', '1', '2', '5'],
 ['3', '5'],
 ['0', '3', '5'],
 ['1', '3', '5'],
 ['0', '1', '3', '5'],
 ['2', '3', '5'],
 ['0', '2', '3', '5'],
 ['1', '2', '3', '5'],
 ['0', '1', '2', '3', '5'],
 ['4', '5'],
 ['0', '4', '5'],
 ['1', '4', '5'],
 ['0', '1', '4', '5'],
 ['2', '4', '5'],
 ['0', '2', '4', '5'],
 ['1', '2', '4', '5'],
 ['0', '1', '2', '4', '5'],
 ['3', '4', '5'],
 ['0', '3', '4', '5'],
 ['1', '3

## Recursive
- Fibonacci sequence
    - Iterative is O(n), recursive without memorizing O(2^n)

In [72]:
#Recursive: O(2^n)
def fib(n):
    if (isinstance(n,int)==False): return None
    if n<0: return None
    if n<=2: return 1
    return fib(n-1)+fib(n-2)

#Iterative: O(n)
def fib_ite(n):
    if (isinstance(n,int)==False): return None
    if n<0: return None
    if n<=2: return 1
    res = [1,1]
    for i in range(n-2):
        res.append(res[-2]+res[-1])
    return res[-1]

#Memorizing - Dynamic Programming
M = {1:1,2:1}
def fib_memo(n):
    if n not in M.keys():
        M[n] = fib_memo(n-1)+fib_memo(n-2)
    return M[n]

In [80]:
del M

In [116]:
%%time
fib(10)

Wall time: 0 ns


55

In [121]:
%%time
fib(30)

Wall time: 840 ms


832040

In [133]:
%%time
fib_ite(100)

Wall time: 0 ns


354224848179261915075

In [74]:
%%time
fib_memo(100)

Wall time: 0 ns


354224848179261915075

In [126]:
def palindrome(s):
    if len(s)<=1: return True
    if s[0]==s[-1]: return palindrome(s[1:-1])
    return False    

In [127]:
%%time
print(palindrome('abcba'))
print(palindrome('abcb'))

True
False
Wall time: 998 µs


## Back Track

In [10]:
class board:
    def __init__(self):
        self.state = [[0]*8 for _ in range(8)]
    
    def place_queen(self,position):
        self.state[position[0]][position[1]]=2
        self.attack_queen()
    
    def remove_queen(self,position):
        if self.state[position[0]][position[1]]!=2: pass
        self.state[position[0]][position[1]]=0
        self.attack_queen()
            
    def attack_queen(self):
        for i in range(len(self.state)):
            for j in range(len(self.state[0])):
                if self.state[i][j]!=2: self.state[i][j]=0
                    
        for i in range(len(self.state)):
            for j in range(len(self.state[0])):
                if self.state[i][j]==2:
                    upper = range(i-1,-1,-1)
                    lower = range(i+1,len(self.state),1)
                    left = range(j-1,-1,-1)
                    right = range(j+1,len(self.state[0]),1)
                    
                    if min(len(upper),len(left))>0:
                        for k in range(min(len(upper),len(left))): 
                            if self.state[upper[k]][left[k]]==0: self.state[upper[k]][left[k]]=1
                    if min(len(upper),len(right))>0:
                        for k in range(min(len(upper),len(right))): 
                            if self.state[upper[k]][right[k]]==0: self.state[upper[k]][right[k]]=1
                    if min(len(lower),len(left))>0:
                        for k in range(min(len(lower),len(left))): 
                            if self.state[lower[k]][left[k]]==0: self.state[lower[k]][left[k]]=1
                    if min(len(lower),len(right))>0:
                        for k in range(min(len(lower),len(right))): 
                            if self.state[lower[k]][right[k]]==0: self.state[lower[k]][right[k]]=1   
                                
                    if len(upper)>0:
                        for k in range(len(upper)):
                            if self.state[upper[k]][j]==0: self.state[upper[k]][j]=1
                    if len(lower)>0:
                        for k in range(len(lower)):
                            if self.state[lower[k]][j]==0: self.state[lower[k]][j]=1       
                    if len(left)>0:
                        for k in range(len(left)):
                            if self.state[i][left[k]]==0: self.state[i][left[k]]=1
                    if len(right)>0:
                        for k in range(len(right)):
                            if self.state[i][right[k]]==0: self.state[i][right[k]]=1   
    
    def count_queen(self):
        cnt = 0
        for i in range(len(self.state)):
            for j in range(len(self.state[0])):
                if self.state[i][j]==2: cnt += 1
        return cnt

In [15]:
b = board()

In [17]:
def queens(b):
    if b.count_queen()==8: 
        return True
    
    for i in range(len(b.state)):
        for j in range(len(b.state[0])):
            if b.state[i][j]==0:
                b.place_queen([i,j])
                if queens(b)==True: 
                    return True
                b.remove_queen([i,j])
    return False                

In [18]:
%%time
queens(b)

Wall time: 5.59 s


True

In [19]:
b.state

[[2, 1, 1, 1, 1, 1, 1, 1],
 [1, 1, 1, 1, 2, 1, 1, 1],
 [1, 1, 1, 1, 1, 1, 1, 2],
 [1, 1, 1, 1, 1, 2, 1, 1],
 [1, 1, 2, 1, 1, 1, 1, 1],
 [1, 1, 1, 1, 1, 1, 2, 1],
 [1, 2, 1, 1, 1, 1, 1, 1],
 [1, 1, 1, 2, 1, 1, 1, 1]]

## Knapsack Problem
- Greedy
- Divide and Conquer
- Dynamic Programming (Memorizing)
- Branch and Bound

In [92]:
def create_itemlist(num):
    names = [chr(97+i) for i in range(num)]
    values = np.random.uniform(5,50,num)
    weights = np.random.uniform(5,20,num)
    items = {}
    for i in range(len(names)):
        item = {}
        item['value'] = int(values[i])
        item['weight'] = int(weights[i])
        items[names[i]] = item
    return items

In [93]:
np.random.seed(1)
items = create_itemlist(10)

In [94]:
items

{'a': {'value': 23, 'weight': 11},
 'b': {'value': 37, 'weight': 15},
 'c': {'value': 5, 'weight': 8},
 'd': {'value': 18, 'weight': 18},
 'e': {'value': 11, 'weight': 5},
 'f': {'value': 9, 'weight': 15},
 'g': {'value': 13, 'weight': 11},
 'h': {'value': 20, 'weight': 13},
 'i': {'value': 22, 'weight': 7},
 'j': {'value': 29, 'weight': 7}}

In [95]:
np.random.seed(2)
items_l = create_itemlist(25)

In [96]:
items_l

{'a': {'value': 24, 'weight': 13},
 'b': {'value': 6, 'weight': 8},
 'c': {'value': 29, 'weight': 6},
 'd': {'value': 24, 'weight': 8},
 'e': {'value': 23, 'weight': 10},
 'f': {'value': 19, 'weight': 12},
 'g': {'value': 14, 'weight': 8},
 'h': {'value': 32, 'weight': 14},
 'i': {'value': 18, 'weight': 12},
 'j': {'value': 17, 'weight': 12},
 'k': {'value': 32, 'weight': 10},
 'l': {'value': 28, 'weight': 16},
 'm': {'value': 11, 'weight': 13},
 'n': {'value': 28, 'weight': 7},
 'o': {'value': 13, 'weight': 15},
 'p': {'value': 40, 'weight': 19},
 'q': {'value': 43, 'weight': 12},
 'r': {'value': 27, 'weight': 18},
 's': {'value': 43, 'weight': 10},
 't': {'value': 8, 'weight': 13},
 'u': {'value': 27, 'weight': 11},
 'v': {'value': 7, 'weight': 11},
 'w': {'value': 24, 'weight': 16},
 'x': {'value': 9, 'weight': 13},
 'y': {'value': 10, 'weight': 19}}

### Greedy
- Computation complexity: O(n)
- But not the global optimum, approximation approach

In [125]:
def greedy_knapsack(items,max_weight):
    bag_weight = 0
    bag_items = []
    for item in sorted(items, key=lambda x: (items[x]['value']), reverse=True):
        if max_weight >= bag_weight+items[item]['weight']:
            bag_weight += items[item]['weight']
            bag_items += [item]
    return sorted(bag_items)

In [126]:
%%time
res = greedy_knapsack(items,80)
print('Items in knapsack:',res)
print('Final value:',sum(items[x]['value'] for x in res))
print('Final weight:',sum(items[x]['weight'] for x in res))

Items in knapsack: ['a', 'b', 'd', 'e', 'h', 'i', 'j']
Final value: 160
Final weight: 76
Wall time: 0 ns


In [127]:
%%time
res = greedy_knapsack(items_l,84)
print('Items in knapsack:',res)
print('Final value:',sum(items_l[x]['value'] for x in res))
print('Final weight:',sum(items_l[x]['weight'] for x in res))

Items in knapsack: ['c', 'h', 'k', 'n', 'p', 'q', 's']
Final value: 247
Final weight: 78
Wall time: 0 ns


### Divide and Conquer - Recursive
- Computation Complexity: O(nW)
- Divide problem to two: 1. a problem when choosing an item A, 2. a problem when removing an item A from the list.

In [154]:
def dc_knapsack2(items,max_weight): #Return: total values, total weights, selected items
    if len(items)==0: 
        return 0,0,[]
    if max_weight < min([items[x]['weight'] for x in items.keys()]):
        return 0,0,[]
    items_removed = {k:v for k,v in items.items()}
    items_removed.pop(list(items.keys())[0])
    
    v1, w1, it1 = dc_knapsack2(items_removed,max_weight)
    v2, w2, it2 = dc_knapsack2(items_removed,max_weight-items[list(items.keys())[0]]['weight'])
    v2 += items[list(items.keys())[0]]['value']
    w2 += items[list(items.keys())[0]]['weight']
    it2 += [list(items.keys())[0]]
    
    if v1>v2:
        return v1,w1,sorted(it1)
    return v2,w2,sorted(it2)

In [152]:
%%time
dc_knapsack2(items,80)

Wall time: 3.99 ms


(160, 77, ['a', 'b', 'c', 'e', 'g', 'h', 'i', 'j'])

In [168]:
%%time
dc_knapsack2(items_l,84)

Wall time: 2.4 s


(266, 83, ['c', 'd', 'k', 'n', 'p', 'q', 's', 'u'])

### Dynamic Programming - Maybe wrong logic somewhere

In [171]:
resdic = {}

def dp_knapsack(items,max_weight): #Return: total values, total weights, selected items
    global resdic
      
    if ( len(items)==0  or  max_weight < min([items[x]['weight'] for x in items.keys()]) ): 
        if (frozenset(items),max_weight) not in resdic.keys():
            resdic[(frozenset(items),max_weight)] = (0,0,[])
        return 0,0,[]

    items_removed = {k:v for k,v in items.items()}
    items_removed.pop(list(items.keys())[0])
    
    if (frozenset(items_removed),max_weight) not in resdic.keys():
        resdic[(frozenset(items_removed),max_weight)] = (dc_knapsack2(items_removed,max_weight))
    if (frozenset(items_removed),max_weight-items[list(items.keys())[0]]['weight']) not in resdic.keys():
        resdic[(frozenset(items_removed),max_weight-items[list(items.keys())[0]]['weight'])] = \
            (dc_knapsack2(items_removed,max_weight-items[list(items.keys())[0]]['weight']))   
        
    v1, w1, it1 = resdic[(frozenset(items_removed),max_weight)]
    v2, w2, it2 = resdic[(frozenset(items_removed),max_weight-items[list(items.keys())[0]]['weight'])]
    
    v2 += items[list(items.keys())[0]]['value']
    w2 += items[list(items.keys())[0]]['weight']
    it2 += [list(items.keys())[0]]
    
    if v1>v2:
        return v1,w1,sorted(it1)
    return v2,w2,sorted(it2)

In [240]:
%%time
dp_knapsack(items,80)

Wall time: 3.89 ms


(160, 77, ['a', 'b', 'c', 'e', 'g', 'h', 'i', 'j'])

In [172]:
%%time
dp_knapsack(items_l,84)

Wall time: 2.45 s


(266, 83, ['c', 'd', 'k', 'n', 'p', 'q', 's', 'u'])

### Branch and Bound
- Lower bound: Greedy algorithm
- Upper bound: Allow partial items packing -> Pack from higher value-weight ratio

In [211]:
def greedy_knapsack_bb(items,max_weight): #Return: total values, total weights, selected items
    
    bag_value = 0
    bag_weight = 0
    bag_items = []
    for item in sorted(items, key=lambda x: (items[x]['value']), reverse=True):
        if max_weight >= bag_weight+items[item]['weight']:
            bag_value += items[item]['value']
            bag_weight += items[item]['weight']
            bag_items += [item]
    return bag_value, bag_weight, sorted(bag_items)

In [212]:
def powdered_knapsack_bb(items,max_weight): 
    
    bag_value = 0
    bag_weight = 0
    bag_items = []
    for item in sorted(items, key=lambda x: items[x]['value']/items[x]['weight'], reverse=True):
        if max_weight >= bag_weight+items[item]['weight']:
            bag_value += items[item]['value']
            bag_weight += items[item]['weight']
            bag_items += [item]
        elif (max_weight > bag_weight and max_weight < bag_weight+items[item]['weight']):
            bag_value += items[item]['value']/items[item]['weight'] * (max_weight - bag_weight)
            bag_weight = max_weight
            bag_items += [item]
    return bag_value, bag_weight, sorted(bag_items)

In [213]:
greedy_knapsack_bb(items,80)

(160, 76, ['a', 'b', 'd', 'e', 'h', 'i', 'j'])

In [214]:
powdered_knapsack_bb(items,80)

(166.0, 80, ['a', 'b', 'd', 'e', 'g', 'h', 'i', 'j'])

In [246]:
powdered_knapsack_bb(items_l,84)

(271.85714285714283, 84, ['c', 'd', 'e', 'h', 'k', 'n', 'q', 's', 'u'])

In [238]:
def bb_knapsack(items,max_weight): #Return: total values, total weights, selected items
    global glob_lower
    global None_cnt
    global Val_cnt
    
    if len(items)==0: 
        return 0,0,[],0,0
    if max_weight < min([items[x]['weight'] for x in items.keys()]):
        return 0,0,[],0,0
    
    glob_lower, _, _ = greedy_knapsack_bb(items,max_weight)
    
    items_removed = {k:v for k,v in items.items()}
    items_removed.pop(list(items.keys())[0])
    
    upper1, _, _ = powdered_knapsack_bb(items_removed,max_weight)
    upper2, _, _ = powdered_knapsack_bb(items_removed,max_weight-items[list(items.keys())[0]]['weight'])
    upper2 += items[list(items.keys())[0]]['value']
    
    if glob_lower <= upper1:
        v1, w1, it1, _, _ = bb_knapsack(items_removed,max_weight)
        v1l,_,_ = greedy_knapsack_bb(items_removed,max_weight)
        Val_cnt += 1
    else:
        v1, w1, it1 = None,None,None
        v1l = None
        None_cnt += 1
    
    if glob_lower <= upper2:
        v2, w2, it2, _, _ = bb_knapsack(items_removed,max_weight-items[list(items.keys())[0]]['weight'])
        v2 += items[list(items.keys())[0]]['value']
        w2 += items[list(items.keys())[0]]['weight']
        it2 += [list(items.keys())[0]]   
        v2l,_,_ = greedy_knapsack_bb(items_removed,max_weight-items[list(items.keys())[0]]['weight'])
        v2l += items[list(items.keys())[0]]['value']
        Val_cnt += 1
    else:
        v2, w2, it2 = None,None,None
        v2l = None
        None_cnt += 1
    
    if (v1==None):
        if glob_lower < v2l: 
            glob_lower = v2l
        return v2,w2,sorted(it2),Val_cnt,None_cnt
    elif (v2==None):
        if glob_lower < v1l: 
            glob_lower = v1l
        return v1,w1,sorted(it1),Val_cnt,None_cnt
    else:
        if glob_lower < min(v1l,v2l): 
            glob_lower = min(v1l,v2l)        
        if v1>v2:
            return v1,w1,sorted(it1),Val_cnt,None_cnt
        return v2,w2,sorted(it2),Val_cnt,None_cnt

In [241]:
%%time
glob_lower = 0
None_cnt = 0
Val_cnt = 1
print(bb_knapsack(items,80))

(160, 77, ['a', 'b', 'c', 'e', 'g', 'h', 'i', 'j'], 43, 28)
Wall time: 1.29 ms


In [239]:
%%time
glob_lower = 0
None_cnt = 0
Val_cnt = 1
print(bb_knapsack(items_l,84))

(266, 83, ['c', 'd', 'k', 'n', 'p', 'q', 's', 'u'], 36862, 23089)
Wall time: 1.21 s


- B&B reduced the number of recursions to 1/30 (1,097,789 -> 36,862)!

## Tree Data Structure - No implementation made
- Class tree (binary search tree)
- function to balance
- Searches: DFS with stack and BFS with queue
- https://pypi.org/project/anytree/